include/revng/Support/Undirected.h

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/iterator.h"
+
+#include "revng/Support/Assert.h"
+
+namespace llvm {
+
+// TODO: Without the `llvm::concat` implementation limitation described below,
+//       we could offer a default implementation for
+//       `GraphTraits<Undirected<T>>` for any `T` for which you have a
+//       definition of both `GraphTraits<T>` and `GraphTraits<Inverse<T>>`
+/// This class is used as a marker class to tell the graph iterator to treat the
+/// underlying graph as an undirected one
+template<class GraphType>
+struct Undirected {
+  const GraphType &Graph;
+
+  inline Undirected(const GraphType &G) : Graph(G) {}
+};
+
+} // namespace llvm
+
+/// Custom child iterator which iterates over the concatenation of the
+/// successors and predecessors of a node. This is used for the `Undirected`
+/// `GraphTraits` specialization. This is done taking inspiration from the
+/// internals of `llvm::concat` implementation, but overcomes one of its
+/// limitations. The limitation is caused by the fact that internally, the `*`
+/// operator of `llvm::concat` (and more specifically the `getHelper`), first
+/// deferences the internal iterator that is being concatenated, and then takes
+/// and returns its pointer (`&*It`). If the iterator used internally, when
+/// deferenced, returns a temporary object with a limited lifetime, which goes
+/// out of scope before its used by the user of `llvm::concat`, we have a
+/// problem.
+// TODO: An alternative to implementing our custom iterator, would be to provide
+//       our own implementation of `revng::concat`, which does not have the
+//       limitation described above.
+template<typename SuccessorIterator,
+         typename PredecessorIterator,
+         typename ProxyObjectType>
```

Comment:
Can we rename this template argument? Iterators have the concept of "value type", i.e., what you get when you `*`. Maybe we can rename this to `ValueType`?

Reply:
Yeah, sound a lot better https://github.com/revng/revng/blob/542a271c138b3255d8c126850db12d2ae3e761d5/include/revng/Support/Undirected.h#L42